File: 2024 May/Day 4 - 881 Boats to Save People [Medium]/sol.cpp
class Solution {
public:
    int numRescueBoats(vector<int>& people, int limit) {
        sort(people.begin(),people.end());
        int i=0,ans=0;
        int j=people.size(); j-=1;

        while(i<=j){
            if(people[i]+people[j]<=limit) i++;
            j--;
            ans++;
        }

        return ans;
    }
};

// C++
// 42 ms (beats 94.77%)
// 45.44 mb (beats 47.33%)

File: 2024 May/Day 4 - 881 Boats to Save People [Medium]/sol.py
class Solution:
    def numRescueBoats(self, people: List[int], limit: int) -> int:
        people.sort()
        ans = 0
        l = 0
        r = len(people) - 1
        
        while l <= r:
            ans += 1
            if people[l] + people[r] <= limit:
                l += 1
            r -= 1

        return ans
    
# Python    
# 341 ms (beats 86.63%)
# 23.22 mb (beats 91.54%)